// ew-asset-registry-contracts/dist/js/src/wrappedContracts/GeneralFunctions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
function getClientVersion(web3) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            web3.currentProvider.send({
                jsonrpc: '2.0',
                method: 'web3_clientVersion',
                params: [],
                id: 1
            }, (e, r) => {
                if (e)
                    reject(e);
                else
                    resolve(r.result);
            });
        });
    });
}
exports.getClientVersion = getClientVersion;
function replayTransaction(web3, txHash) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            web3.currentProvider.send({
                jsonrpc: '2.0',
                method: 'trace_replayTransaction',
                params: [txHash, ['trace', 'vmTrace', 'stateDiff']],
                id: 1,
            }, (e, r) => {
                if (e)
                    reject(e);
                else
                    resolve(r.result);
            });
        });
    });
}
exports.replayTransaction = replayTransaction;
class GeneralFunctions {
    constructor(web3Contract) {
        this.web3Contract = web3Contract;
    }
    sendRaw(web3, privateKey, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            const txData = {
                nonce: txParams.nonce,
                gasLimit: txParams.gas,
                gasPrice: 0,
                data: txParams.data,
                from: txParams.from,
                to: txParams.to
            };
            const txObject = yield web3.eth.accounts.signTransaction(txData, privateKey);
            return (yield web3.eth.sendSignedTransaction(txObject.rawTransaction));
        });
    }
    getWeb3Contract() {
        return this.web3Contract;
    }
    getErrorMessage(web3, txObj) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new Promise((resolve, reject) => {
                web3.currentProvider.send({
                    jsonrpc: '2.0',
                    method: 'trace_call',
                    params: [txObj, ['trace']],
                    id: 1,
                }, (e, r) => {
                    if (e)
                        reject(e);
                    else {
                        const outputResult = r.result.output;
                        let shorterAsciiCode = '0x' + outputResult.substr(10);
                        if (r.result.output === "0x") {
                            resolve("Bad instruction / revert without reason string");
                        }
                        resolve(web3.utils.toAscii(shorterAsciiCode));
                    }
                });
            });
        });
    }
}
exports.GeneralFunctions = GeneralFunctions;
