// ew-utils-general-lib/dist/js/blockchain-facade/EnumExports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TimeFrame;
(function (TimeFrame) {
    TimeFrame[TimeFrame["yearly"] = 0] = "yearly";
    TimeFrame[TimeFrame["monthly"] = 1] = "monthly";
    TimeFrame[TimeFrame["daily"] = 2] = "daily";
    TimeFrame[TimeFrame["hourly"] = 3] = "hourly";
})(TimeFrame = exports.TimeFrame || (exports.TimeFrame = {}));
var Currency;
(function (Currency) {
    Currency[Currency["Euro"] = 0] = "Euro";
    Currency[Currency["USD"] = 1] = "USD";
    Currency[Currency["SingaporeDollar"] = 2] = "SingaporeDollar";
    Currency[Currency["Ether"] = 3] = "Ether";
})(Currency = exports.Currency || (exports.Currency = {}));
var AssetType;
(function (AssetType) {
    AssetType[AssetType["Wind"] = 0] = "Wind";
    AssetType[AssetType["Solar"] = 1] = "Solar";
    AssetType[AssetType["RunRiverHydro"] = 2] = "RunRiverHydro";
    AssetType[AssetType["BiomassGas"] = 3] = "BiomassGas";
    AssetType[AssetType["Battery"] = 4] = "Battery";
})(AssetType = exports.AssetType || (exports.AssetType = {}));
var Compliance;
(function (Compliance) {
    Compliance[Compliance["none"] = 0] = "none";
    Compliance[Compliance["IREC"] = 1] = "IREC";
    Compliance[Compliance["EEC"] = 2] = "EEC";
    Compliance[Compliance["TIGR"] = 3] = "TIGR";
})(Compliance = exports.Compliance || (exports.Compliance = {}));
