// ew-user-registry-contracts/dist/js/src/index.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Martin Kuechler, martin.kuchler@slock.it; Heiko Burkhardt, heiko.burkhardt@slock.it
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var UserLogic_1 = require("./wrappedContracts/UserLogic");
exports.UserLogic = UserLogic_1.UserLogic;
var UserContractLookup_1 = require("./wrappedContracts/UserContractLookup");
exports.UserContractLookup = UserContractLookup_1.UserContractLookup;
var migrateContracts_1 = require("./utils/migrateContracts");
exports.migrateUserRegistryContracts = migrateContracts_1.migrateUserRegistryContracts;
const UserDB_json_1 = __importDefault(require("../contract-build/UserDB.json"));
exports.UserDBJSON = UserDB_json_1.default;
const UserLogic_json_1 = __importDefault(require("../contract-build/UserLogic.json"));
exports.UserLogicJSON = UserLogic_json_1.default;
const UserContractLookup_json_1 = __importDefault(require("../contract-build/UserContractLookup.json"));
exports.UserContractLookupJSON = UserContractLookup_json_1.default;
