// ew-asset-registry-lib/dist/js/src/blockchain-facade/ProducingAsset.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Heiko Burkhardt, heiko.burkhardt@slock.it; Martin Kuechler, martin.kuchler@slock.it
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Asset = __importStar(require("./Asset"));
const __1 = require("..");
/**
 * TODO: in generalLib
 */
var Type;
(function (Type) {
    Type[Type["Wind"] = 0] = "Wind";
    Type[Type["Solar"] = 1] = "Solar";
    Type[Type["RunRiverHydro"] = 2] = "RunRiverHydro";
    Type[Type["BiomassGas"] = 3] = "BiomassGas";
    Type[Type["Battery"] = 4] = "Battery";
})(Type = exports.Type || (exports.Type = {}));
var Compliance;
(function (Compliance) {
    Compliance[Compliance["none"] = 0] = "none";
    Compliance[Compliance["IREC"] = 1] = "IREC";
    Compliance[Compliance["EEC"] = 2] = "EEC";
    Compliance[Compliance["TIGR"] = 3] = "TIGR";
})(Compliance = exports.Compliance || (exports.Compliance = {}));
exports.getAssetListLength = (configuration) => __awaiter(this, void 0, void 0, function* () {
    return parseInt(yield configuration.blockchainProperties.producingAssetLogicInstance.getAssetListLength(), 10);
});
exports.getAllAssets = (configuration) => __awaiter(this, void 0, void 0, function* () {
    const assetsPromises = Array(yield exports.getAssetListLength(configuration))
        .fill(null)
        .map((item, index) => (new Entity(index.toString(), configuration)).sync());
    return Promise.all(assetsPromises);
});
exports.getAllAssetsOwnedBy = (owner, configuration) => __awaiter(this, void 0, void 0, function* () {
    return (yield exports.getAllAssets(configuration))
        .filter((asset) => asset.owner.address.toLowerCase() === owner.toLowerCase());
});
exports.createAsset = (assetPropertiesOnChain, assetPropertiesOffChain, configuration) => __awaiter(this, void 0, void 0, function* () {
    const producingAsset = new Entity(null, configuration);
    const offChainStorageProperties = producingAsset.prepareEntityCreation(assetPropertiesOnChain, assetPropertiesOffChain, __1.ProducingAssetPropertiesOffchainSchema);
    if (configuration.offChainDataSource) {
        assetPropertiesOnChain.url = producingAsset.getUrl();
        assetPropertiesOnChain.propertiesDocumentHash = offChainStorageProperties.rootHash;
    }
    const tx = yield configuration.blockchainProperties.producingAssetLogicInstance.createSonnenAsset(assetPropertiesOnChain.smartMeter.address, assetPropertiesOnChain.owner.address, assetPropertiesOnChain.active, assetPropertiesOnChain.matcher.map((matcher) => matcher.address), assetPropertiesOnChain.propertiesDocumentHash, assetPropertiesOnChain.url, assetPropertiesOnChain.maxOwnerChanges, assetPropertiesOnChain.marketLookupAddress, {
        from: configuration.blockchainProperties.activeUser.address,
        privateKey: configuration.blockchainProperties.activeUser.privateKey,
    });
    producingAsset.id = configuration.blockchainProperties.web3.utils.hexToNumber(tx.logs[0].topics[1]).toString();
    yield producingAsset.putToOffChainStorage(assetPropertiesOffChain, offChainStorageProperties);
    if (configuration.logger) {
        configuration.logger.info(`Producing asset ${producingAsset.id} created`);
    }
    return producingAsset.sync();
});
class Entity extends Asset.Entity {
    getUrl() {
        return `${this.configuration.offChainDataSource.baseUrl}/ProducingAsset`;
    }
    sync() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.id != null) {
                const asset = yield this.configuration.blockchainProperties.producingAssetLogicInstance.getSonnenAssetById(this.id);
                this.smartMeter = { address: asset.assetGeneral.smartMeter };
                this.owner = { address: asset.assetGeneral.owner };
                this.lastSmartMeterReadWh = asset.assetGeneral.lastSmartMeterReadWh;
                this.active = asset.assetGeneral.active;
                this.lastSmartMeterReadFileHash = asset.assetGeneral.lastSmartMeterReadFileHash;
                this.matcher = [{ address: asset.assetGeneral.matcher }];
                this.propertiesDocumentHash = asset.assetGeneral.propertiesDocumentHash;
                this.url = asset.assetGeneral.url;
                this.initialized = true;
                this.maxOwnerChanges = asset.maxOwnerChanges;
                this.marketLookupAddress = asset.marketProps.marketLookupAddress;
                this.maxOwnerChanges = asset.maxOwnerChanges;
                this.marketLookupAddress = asset.marketProps.marketLookupAddress;
                this.supplyId = asset.marketProps.supplyId;
                this.timeFrameFrom = asset.marketProps.timeFrameFrom;
                this.timeFrameTo = asset.marketProps.timeFrameTo;
                this.averagePower = asset.marketProps.averagePower;
                this.baselinePower = asset.marketProps.baselinePower;
                this.supplyIdSet = asset.marketProps.supplyIdSet;
                this.certificateOwner = asset.marketProps.certificateOwner;
                this.powerProfileURL = asset.marketProps.powerProfileURL;
                this.offChainProperties = yield this.getOffChainProperties(this.propertiesDocumentHash);
                if (this.configuration.logger) {
                    this.configuration.logger.verbose(`Producing asset ${this.id} synced`);
                }
            }
            return this;
        });
    }
    saveSmartMeterRead(meterReading, filehash, timeFrameFrom, timeFrameTo, averagePower, baselinePower, powerProfileURL) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.configuration.blockchainProperties.activeUser.privateKey) {
                return this.configuration.blockchainProperties.producingAssetLogicInstance.saveSonnenSmartMeterRead(this.id, meterReading, filehash, timeFrameFrom, timeFrameTo, averagePower, baselinePower, powerProfileURL, { privateKey: this.configuration.blockchainProperties.activeUser.privateKey });
            }
            else {
                return this.configuration.blockchainProperties.producingAssetLogicInstance.saveSmartMeterRead(this.id, meterReading, filehash, timeFrameFrom, timeFrameTo, averagePower, baselinePower, powerProfileURL, { from: this.configuration.blockchainProperties.activeUser.address });
            }
        });
    }
}
exports.Entity = Entity;
