// ew-market-contracts/dist/js/src/index.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Martin Kuechler, martin.kuchler@slock.it; Heiko Burkhardt, heiko.burkhardt@slock.it
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const MarketContractLookup_json_1 = __importDefault(require("../contract-build/MarketContractLookup.json"));
exports.MarketContractLookupJSON = MarketContractLookup_json_1.default;
const MarketDB_json_1 = __importDefault(require("../contract-build/MarketDB.json"));
exports.MarketDBJSON = MarketDB_json_1.default;
const MarketLogic_json_1 = __importDefault(require("../contract-build/MarketLogic.json"));
exports.MarketLogicJSON = MarketLogic_json_1.default;
var MarketLogic_1 = require("./wrappedContracts/MarketLogic");
exports.MarketLogic = MarketLogic_1.MarketLogic;
var MarketContractLookup_1 = require("./wrappedContracts/MarketContractLookup");
exports.MarketContractLookup = MarketContractLookup_1.MarketContractLookup;
var migrateContracts_1 = require("./utils/migrateContracts");
exports.migrateMarketRegistryContracts = migrateContracts_1.migrateMarketRegistryContracts;
