// ew-origin-contracts/dist/js/src/index.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Martin Kuechler, martin.kuchler@slock.it; Heiko Burkhardt, heiko.burkhardt@slock.it;
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var migrateContracts_1 = require("./utils/migrateContracts");
exports.migrateCertificateRegistryContracts = migrateContracts_1.migrateCertificateRegistryContracts;
exports.migrateEnergyBundleContracts = migrateContracts_1.migrateEnergyBundleContracts;
exports.migrateSonnenContracts = migrateContracts_1.migrateSonnenContracts;
var CertificateLogic_1 = require("./wrappedContracts/CertificateLogic");
exports.CertificateLogic = CertificateLogic_1.CertificateLogic;
var EnergyCertificateBundleLogic_1 = require("./wrappedContracts/EnergyCertificateBundleLogic");
exports.EnergyCertificateBundleLogic = EnergyCertificateBundleLogic_1.EnergyCertificateBundleLogic;
var EnergyLogic_1 = require("./wrappedContracts/EnergyLogic");
exports.EnergyLogic = EnergyLogic_1.EnergyLogic;
var OriginContractLookup_1 = require("./wrappedContracts/OriginContractLookup");
exports.OriginContractLookup = OriginContractLookup_1.OriginContractLookup;
const CertificateDB_json_1 = __importDefault(require("../contract-build/CertificateDB.json"));
exports.CertificateDBJSON = CertificateDB_json_1.default;
const CertificateLogic_json_1 = __importDefault(require("../contract-build/CertificateLogic.json"));
exports.CertificateLogicJSON = CertificateLogic_json_1.default;
const CertificateSpecificContract_json_1 = __importDefault(require("../contract-build/CertificateSpecificContract.json"));
exports.CertificateSpecificContractJSON = CertificateSpecificContract_json_1.default;
const CertificateSpecificDB_json_1 = __importDefault(require("../contract-build/CertificateSpecificDB.json"));
exports.CertificateSpecificDBJSON = CertificateSpecificDB_json_1.default;
const EnergyCertificateBundleDB_json_1 = __importDefault(require("../contract-build/EnergyCertificateBundleDB.json"));
exports.EnergyCertificateBundleDBJSON = EnergyCertificateBundleDB_json_1.default;
const EnergyCertificateBundleLogic_json_1 = __importDefault(require("../contract-build/EnergyCertificateBundleLogic.json"));
exports.EnergyCertificateBundleLogicJSON = EnergyCertificateBundleLogic_json_1.default;
const EnergyDB_json_1 = __importDefault(require("../contract-build/EnergyDB.json"));
exports.EnergyDBJSON = EnergyDB_json_1.default;
const EnergyLogic_json_1 = __importDefault(require("../contract-build/EnergyLogic.json"));
exports.EnergyLogicJSON = EnergyLogic_json_1.default;
const OriginContractLookup_json_1 = __importDefault(require("../contract-build/OriginContractLookup.json"));
exports.OriginContractLookupJSON = OriginContractLookup_json_1.default;
const TradableEntityContract_json_1 = __importDefault(require("../contract-build/TradableEntityContract.json"));
exports.TradableEntityContractJSON = TradableEntityContract_json_1.default;
const TradableEntityDB_json_1 = __importDefault(require("../contract-build/TradableEntityDB.json"));
exports.TradableEntityDBJSON = TradableEntityDB_json_1.default;
const TradableEntityLogic_json_1 = __importDefault(require("../contract-build/TradableEntityLogic.json"));
exports.TradableEntityLogicJSON = TradableEntityLogic_json_1.default;
