// ew-origin-contracts/dist/js/src/wrappedContracts/EnergyCertificateBundleDB.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const GeneralFunctions_1 = require("./GeneralFunctions");
const EnergyCertificateBundleDB_json_1 = __importDefault(require("../../contract-build/EnergyCertificateBundleDB.json"));
class EnergyCertificateBundleDB extends GeneralFunctions_1.GeneralFunctions {
    constructor(web3, address) {
        super(address ? new web3.eth.Contract(EnergyCertificateBundleDB_json_1.default.abi, address) : new web3.eth.Contract(EnergyCertificateBundleDB_json_1.default.abi, EnergyCertificateBundleDB_json_1.default.networks.length > 0 ? (EnergyCertificateBundleDB_json_1.default.networks[0]) : null));
        this.buildFile = EnergyCertificateBundleDB_json_1.default;
        this.web3 = web3;
    }
    getAllLogChangeOwnerEvents(eventFilter) {
        return __awaiter(this, void 0, void 0, function* () {
            let filterParams;
            if (eventFilter) {
                filterParams = {
                    fromBlock: eventFilter.fromBlock ? eventFilter.fromBlock : 0,
                    toBlock: eventFilter.toBlock ? eventFilter.toBlock : 'latest'
                };
                if (eventFilter.topics) {
                    filterParams.topics = eventFilter.topics;
                }
            }
            else {
                filterParams = {
                    fromBlock: 0,
                    toBlock: 'latest'
                };
            }
            return yield this.web3Contract.getPastEvents('LogChangeOwner', filterParams);
        });
    }
    getAllEvents(eventFilter) {
        return __awaiter(this, void 0, void 0, function* () {
            let filterParams;
            if (eventFilter) {
                filterParams = {
                    fromBlock: eventFilter.fromBlock ? eventFilter.fromBlock : 0,
                    toBlock: eventFilter.toBlock ? eventFilter.toBlock : 'latest',
                    topics: eventFilter.topics ? eventFilter.topics : [null]
                };
            }
            else {
                filterParams = {
                    fromBlock: 0,
                    toBlock: 'latest',
                    topics: [null]
                };
            }
            return yield this.web3Contract.getPastEvents('allEvents', filterParams);
        });
    }
    getTradableEntityOwner(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntityOwner(_entityId).call(txParams));
        });
    }
    getApproved(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getApproved(_entityId).call(txParams));
        });
    }
    addEscrowForEntity(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntityEscrowExternal(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getTradableEntityEscrowLength(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntityEscrowLength(_entityId).call(txParams));
        });
    }
    getBundleListLength(txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getBundleListLength().call(txParams));
        });
    }
    setDataLog(_certificateId, _newDataLog, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setDataLog(_certificateId, _newDataLog)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setDataLog(_certificateId, _newDataLog)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setDataLog(_certificateId, _newDataLog)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getOwnerChangeCounter(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.getOwnerChangeCounter(_certificateId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.getOwnerChangeCounter(_certificateId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.getOwnerChangeCounter(_certificateId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getBundle(_bundleID, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getBundle(_bundleID).call(txParams));
        });
    }
    getRetired(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.getRetired(_certificateId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.getRetired(_certificateId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.getRetired(_certificateId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setOwnerChangeCounter(_certificateId, _newOwnerChangeCounter, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setOwnerChangeCounter(_certificateId, _newOwnerChangeCounter)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setOwnerChangeCounter(_certificateId, _newOwnerChangeCounter)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setOwnerChangeCounter(_certificateId, _newOwnerChangeCounter)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    addApprovalExternal(_entityId, _approve, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setOnChainDirectPurchasePrice(_entityId, _price, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getOwnerToOperators(_company, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getOwnerToOperators(_company, _escrow).call(txParams));
        });
    }
    setTradableEntityOwnerExternal(_entityId, _owner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getTradableEntity(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntity(_entityId).call(txParams));
        });
    }
    addChildrenExternal(_certificateId, _childId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addChildrenExternal(_certificateId, _childId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addChildrenExternal(_certificateId, _childId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addChildrenExternal(_certificateId, _childId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    owner(txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.owner().call(txParams));
        });
    }
    getCertificateSpecific(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getCertificateSpecific(_certificateId).call(txParams));
        });
    }
    getBalanceOf(_owner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getBalanceOf(_owner).call(txParams));
        });
    }
    setRetired(_certificateId, _retired, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setRetired(_certificateId, _retired)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setRetired(_certificateId, _retired)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setRetired(_certificateId, _retired)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setMaxOwnerChanges(_certificateId, _newMaxOwnerChanges, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setMaxOwnerChanges(_certificateId, _newMaxOwnerChanges)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setMaxOwnerChanges(_certificateId, _newMaxOwnerChanges)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setMaxOwnerChanges(_certificateId, _newMaxOwnerChanges)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    changeOwner(_newOwner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.changeOwner(_newOwner)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.changeOwner(_newOwner)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.changeOwner(_newOwner)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getMaxOwnerChanges(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.getMaxOwnerChanges(_certificateId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.getMaxOwnerChanges(_certificateId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.getMaxOwnerChanges(_certificateId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    removeTokenAndPrice(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    addChildren(_certificateId, _childId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addChildren(_certificateId, _childId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addChildren(_certificateId, _childId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addChildren(_certificateId, _childId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setCertificateSpecific(_certificateId, _certificate, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setCertificateSpecific(_certificateId, _certificate)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setCertificateSpecific(_certificateId, _certificate)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setCertificateSpecific(_certificateId, _certificate)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setOwnerToOperators(_company, _escrow, _allowed, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getCertificateChildrenLength(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getCertificateChildrenLength(_certificateId).call(txParams));
        });
    }
    removeEscrow(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntity(_entityId, _entity, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableToken(_entityId, _token, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getOnChainDirectPurchasePrice(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getOnChainDirectPurchasePrice(_entityId).call(txParams));
        });
    }
    getDataLog(_certificateId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.getDataLog(_certificateId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.getDataLog(_certificateId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.getDataLog(_certificateId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getTradableToken(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableToken(_entityId).call(txParams));
        });
    }
    createEnergyCertificateBundle(_tradableEntity, _certificateSpecific, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.createEnergyCertificateBundle(_tradableEntity, _certificateSpecific)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.createEnergyCertificateBundle(_tradableEntity, _certificateSpecific)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.createEnergyCertificateBundle(_tradableEntity, _certificateSpecific)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
}
exports.EnergyCertificateBundleDB = EnergyCertificateBundleDB;
