// ew-origin-contracts/dist/js/src/wrappedContracts/EnergyDB.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const GeneralFunctions_1 = require("./GeneralFunctions");
const EnergyDB_json_1 = __importDefault(require("../../contract-build/EnergyDB.json"));
class EnergyDB extends GeneralFunctions_1.GeneralFunctions {
    constructor(web3, address) {
        super(address ? new web3.eth.Contract(EnergyDB_json_1.default.abi, address) : new web3.eth.Contract(EnergyDB_json_1.default.abi, EnergyDB_json_1.default.networks.length > 0 ? (EnergyDB_json_1.default.networks[0]) : null));
        this.buildFile = EnergyDB_json_1.default;
        this.web3 = web3;
    }
    getAllLogChangeOwnerEvents(eventFilter) {
        return __awaiter(this, void 0, void 0, function* () {
            let filterParams;
            if (eventFilter) {
                filterParams = {
                    fromBlock: eventFilter.fromBlock ? eventFilter.fromBlock : 0,
                    toBlock: eventFilter.toBlock ? eventFilter.toBlock : 'latest'
                };
                if (eventFilter.topics) {
                    filterParams.topics = eventFilter.topics;
                }
            }
            else {
                filterParams = {
                    fromBlock: 0,
                    toBlock: 'latest'
                };
            }
            return yield this.web3Contract.getPastEvents('LogChangeOwner', filterParams);
        });
    }
    getAllEvents(eventFilter) {
        return __awaiter(this, void 0, void 0, function* () {
            let filterParams;
            if (eventFilter) {
                filterParams = {
                    fromBlock: eventFilter.fromBlock ? eventFilter.fromBlock : 0,
                    toBlock: eventFilter.toBlock ? eventFilter.toBlock : 'latest',
                    topics: eventFilter.topics ? eventFilter.topics : [null]
                };
            }
            else {
                filterParams = {
                    fromBlock: 0,
                    toBlock: 'latest',
                    topics: [null]
                };
            }
            return yield this.web3Contract.getPastEvents('allEvents', filterParams);
        });
    }
    getTradableEntityOwner(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntityOwner(_entityId).call(txParams));
        });
    }
    getApproved(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getApproved(_entityId).call(txParams));
        });
    }
    addEscrowForEntity(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addEscrowForEntity(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntityEscrowExternal(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityEscrowExternal(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getTradableEntityEscrowLength(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntityEscrowLength(_entityId).call(txParams));
        });
    }
    addEscrowForAsset(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addEscrowForAsset(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addEscrowForAsset(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addEscrowForAsset(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setApproved(_entityId, _approved, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setApproved(_entityId, _approved)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setApproved(_entityId, _approved)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setApproved(_entityId, _approved)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityOwnerAndAddApproval(_entityId, _owner, _approve)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    addApprovalExternal(_entityId, _approve, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.addApprovalExternal(_entityId, _approve)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setOnChainDirectPurchasePrice(_entityId, _price, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setOnChainDirectPurchasePrice(_entityId, _price)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    createTradableEntityEntry(_assetId, _owner, _powerInW, _flex, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.createTradableEntityEntry(_assetId, _owner, _powerInW, _flex)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.createTradableEntityEntry(_assetId, _owner, _powerInW, _flex)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.createTradableEntityEntry(_assetId, _owner, _powerInW, _flex)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setEscrow(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setEscrow(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setEscrow(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setEscrow(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getOwnerToOperators(_company, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getOwnerToOperators(_company, _escrow).call(txParams));
        });
    }
    setFexibility(_entityId, _flex, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setFexibility(_entityId, _flex)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setFexibility(_entityId, _flex)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setFexibility(_entityId, _flex)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntityOwnerExternal(_entityId, _owner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntityOwnerExternal(_entityId, _owner)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getTradableEntity(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableEntity(_entityId).call(txParams));
        });
    }
    owner(txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.owner().call(txParams));
        });
    }
    getBalanceOf(_owner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getBalanceOf(_owner).call(txParams));
        });
    }
    setFileHash(_entityId, _fileHash, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setFileHash(_entityId, _fileHash)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setFileHash(_entityId, _fileHash)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setFileHash(_entityId, _fileHash)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    changeOwner(_newOwner, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.changeOwner(_newOwner)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.changeOwner(_newOwner)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.changeOwner(_newOwner)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getCertificateListLength(txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getCertificateListLength().call(txParams));
        });
    }
    removeTokenAndPrice(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.removeTokenAndPrice(_entityId)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setOwnerToOperators(_company, _escrow, _allowed, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setOwnerToOperators(_company, _escrow, _allowed)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getEnergyCertificate(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getEnergyCertificate(_entityId).call(txParams));
        });
    }
    removeEscrow(_entityId, _escrow, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.removeEscrow(_entityId, _escrow)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableEntity(_entityId, _entity, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableEntity(_entityId, _entity)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    setTradableToken(_entityId, _token, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            let transactionParams;
            const txData = yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                .encodeABI();
            let gas;
            if (txParams) {
                if (txParams.privateKey) {
                    const privateKey = txParams.privateKey.startsWith("0x") ? txParams.privateKey : "0x" + txParams.privateKey;
                    txParams.from = this.web3.eth.accounts.privateKeyToAccount(privateKey).address;
                    txParams.nonce = txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from));
                }
                if (!txParams.gas) {
                    try {
                        gas = yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                            .estimateGas({ from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0] });
                    }
                    catch (ex) {
                        if (!(yield GeneralFunctions_1.getClientVersion(this.web3)).includes('Parity'))
                            throw new Error(ex);
                        const errorResult = yield this.getErrorMessage(this.web3, {
                            from: txParams ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                            to: this.web3Contract._address,
                            data: txData,
                            gas: this.web3.utils.toHex(7000000)
                        });
                        throw new Error(errorResult);
                    }
                    gas = Math.round(gas * 2);
                    txParams.gas = gas;
                }
                transactionParams = {
                    from: txParams.from ? txParams.from : (yield this.web3.eth.getAccounts())[0],
                    gas: txParams.gas ? txParams.gas : Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: txParams.nonce ? txParams.nonce : (yield this.web3.eth.getTransactionCount(txParams.from)),
                    data: txParams.data ? txParams.data : '',
                    to: this.web3Contract._address,
                    privateKey: txParams.privateKey ? txParams.privateKey : ""
                };
            }
            else {
                transactionParams = { from: (yield this.web3.eth.getAccounts())[0],
                    gas: Math.round(gas * 1.1 + 21000),
                    gasPrice: 0,
                    nonce: (yield this.web3.eth.getTransactionCount((yield this.web3.eth.getAccounts())[0])),
                    data: '',
                    to: this.web3Contract._address,
                    privateKey: ""
                };
            }
            if (transactionParams.privateKey !== '') {
                transactionParams.data = txData;
                return (yield this.sendRaw(this.web3, transactionParams.privateKey, transactionParams));
            }
            else {
                return yield this.web3Contract.methods.setTradableToken(_entityId, _token)
                    .send({ from: transactionParams.from, gas: transactionParams.gas });
            }
        });
    }
    getOnChainDirectPurchasePrice(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getOnChainDirectPurchasePrice(_entityId).call(txParams));
        });
    }
    getTradableToken(_entityId, txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.web3Contract.methods.getTradableToken(_entityId).call(txParams));
        });
    }
}
exports.EnergyDB = EnergyDB;
