// ew-asset-registry-contracts/dist/js/src/index.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Martin Kuechler, martin.kuchler@slock.it; Heiko Burkhardt, heiko.burkhardt@slock.it
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var AssetConsumingRegistryLogic_1 = require("./wrappedContracts/AssetConsumingRegistryLogic");
exports.AssetConsumingRegistryLogic = AssetConsumingRegistryLogic_1.AssetConsumingRegistryLogic;
var AssetProducingRegistryLogic_1 = require("./wrappedContracts/AssetProducingRegistryLogic");
exports.AssetProducingRegistryLogic = AssetProducingRegistryLogic_1.AssetProducingRegistryLogic;
var AssetContractLookup_1 = require("./wrappedContracts/AssetContractLookup");
exports.AssetContractLookup = AssetContractLookup_1.AssetContractLookup;
var migrateContracts_1 = require("./utils/migrateContracts");
exports.migrateAssetRegistryContracts = migrateContracts_1.migrateAssetRegistryContracts;
exports.migrateSonnenAssetRegistryContracts = migrateContracts_1.migrateSonnenAssetRegistryContracts;
const AssetConsumingDB_json_1 = __importDefault(require("../contract-build/AssetConsumingDB.json"));
exports.AssetConsumingDBJSON = AssetConsumingDB_json_1.default;
const AssetConsumingRegistryLogic_json_1 = __importDefault(require("../contract-build/AssetConsumingRegistryLogic.json"));
exports.AssetConsumingRegistryLogicJSON = AssetConsumingRegistryLogic_json_1.default;
const AssetProducingDB_json_1 = __importDefault(require("../contract-build/AssetProducingDB.json"));
exports.AssetProducingDBJSON = AssetProducingDB_json_1.default;
const AssetProducingRegistryLogic_json_1 = __importDefault(require("../contract-build/AssetProducingRegistryLogic.json"));
exports.AssetProducingRegistryLogicJSON = AssetProducingRegistryLogic_json_1.default;
const AssetContractLookup_json_1 = __importDefault(require("../contract-build/AssetContractLookup.json"));
exports.AssetContractLookupJSON = AssetContractLookup_json_1.default;
const SonnenProducingAssetLogic_json_1 = __importDefault(require("../contract-build/SonnenProducingAssetLogic.json"));
exports.SonnenAssetProducingRegistryLogicJSON = SonnenProducingAssetLogic_json_1.default;
const SonnenAssetProducingDB_json_1 = __importDefault(require("../contract-build/SonnenAssetProducingDB.json"));
exports.SonnenAssetProducingDBJSON = SonnenAssetProducingDB_json_1.default;
var SonnenProducingAssetLogic_1 = require("./wrappedContracts/SonnenProducingAssetLogic");
exports.SonnenProducingAssetLogic = SonnenProducingAssetLogic_1.SonnenProducingAssetLogic;
