// ew-asset-registry-lib/dist/js/src/index.js
"use strict";
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Heiko Burkhardt, heiko.burkhardt@slock.it; Martin Kuechler, martin.kuchler@slock.it
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ProducingAsset = __importStar(require("./blockchain-facade/ProducingAsset"));
exports.ProducingAsset = ProducingAsset;
const ConsumingAsset = __importStar(require("./blockchain-facade/ConsumingAsset"));
exports.ConsumingAsset = ConsumingAsset;
const Asset = __importStar(require("./blockchain-facade/Asset"));
exports.Asset = Asset;
const AssetPropertiesOffChain_schema_json_1 = __importDefault(require("../schemas/AssetPropertiesOffChain.schema.json"));
exports.AssetPropertiesOffchainSchema = AssetPropertiesOffChain_schema_json_1.default;
const ProducingAssetPropertiesOffChain_schema_json_1 = __importDefault(require("../schemas/ProducingAssetPropertiesOffChain.schema.json"));
exports.ProducingAssetPropertiesOffchainSchema = ProducingAssetPropertiesOffChain_schema_json_1.default;
const BlockchainPropertiesFactory_1 = require("./blockchain-facade/BlockchainPropertiesFactory");
exports.createBlockchainProperties = BlockchainPropertiesFactory_1.createBlockchainProperties;
